### ISSUE TITLE
Ring Attention w/ document packing produces different results

### ISSUE THREAD

**Issue by Ueaj-Kerman** [axolotl-ai-cloud/axolotl#3608]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

When using document packing, ring attention should produce similar or bitwise identical outputs to varlen_llama3 or padded batches.

### Current behaviour

When using specifically document packing with ring attention, the loss on a given step is much higher and different than would otherwise be expected. Ring attention on a single non-packed document works. Document packing with varlen_llama3 attention works.

**Multi-doc packed data** (`data/packed.pt`, pack 0 = 5 docs with
position_id resets at each boundary):

| config                                | step-0 loss | ratio to cp=1      |
|---------------------------------------|-------------|--------------------|
| `cp=1` (baseline)                     | `3.171875`  | 1.000×             |
| `cp=2, ring_attn_func=batch_ring`     | `5.843750`  | **1.843×** (bug)   |
| `cp=2, ring_attn_func=varlen_llama3`  | `3.156250`  | 0.995× (fix)       |

**Single-doc control** (`data/packed_single.pt`, one 1024-token doc,
`position_ids` 0..1023 contiguous, zero resets). Every step is
bitwise-identical across all three configs:

| config                                | loss (all steps) |
|---------------------------------------|------------------|
| `cp=1` (baseline)                     | `3.031250`       |
| `cp=2, ring_attn_func=batch_ring`     | `3.031250`       |
| `cp=2, ring_attn_func=varlen_llama3`  | `3.031250`       |

### Steps to reproduce

I don't have the skill set to debug exactly what is wrong with it, but Claude was able to produce a minimal reproduction of the issue which can be found here: https://github.com/Ueaj-Kerman/axolotl-cp-repro

### Config yaml

```yaml

```

### Possible solution

I'd bet the document ids aren't being rotated along the ring, causing the issue

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11.15

### axolotl branch-commit

08fc7de87e79f38c367f6776c5111b40a914062e

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hey, thanks for the report and repro. Just wanted to check one thing, in axolotl, we set to varlen if CP and sample packing, which according to your metrics work?

For CP no packing, I believe it defaults to batch. In which case do you encounter it setting batch ring for CP+sample packing?

**Comment by Ueaj-Kerman** [comment]
It works if there's only one sample in the pack, but if there are multiple, it breaks

**Comment by NanoCode012** [comment]
> It works if there's only one sample in the pack, but if there are multiple, it breaks

Per your results, "5 docs cp=2, ring_attn_func=varlen_llama3" has close loss to baseline. Is this not correct?

Could you run a axolotl e2e test with wandb metrics to demonstrate the issue?

**Comment by reallyticsai** [comment]
Looks like you're hitting a significant discrepancy with the loss values when using ring attention with document packing. From my experience, issues like this can sometimes stem from how attention masks are applied or how the position IDs are handled in packed batches. 

Here are a couple of thoughts that might help debug this:

- **Check Position ID Handling**: Ensure that the position IDs do reset correctly at each boundary when packing documents. If they're not being reset properly, it could lead to the model misinterpreting the context of each document.
  
- **Compare Attention Masking**: Investigate how the attention masks are constructed for both the ring attention and the varlen_llama3 configurations. Differences in attention masking can lead to variations in loss.

- **Experiment with Hyperparameters**: If you're locked to using ring attention, tweak the hyperparameters slightly to see if there's a sensitivity in how the model is learning from the packed data.

If you get a chance, try logging the intermediate outputs for both configurations as well. Sometimes the actual attention weights can reveal discrepancies that the loss alone won't show.

**Comment by smqd19** [comment]
I've seen similar issues in production when dealing with attention mechanisms and packed documents. Here are a few things to check:

- **Position IDs**: Ensure that position IDs are correctly reset at document boundaries. Incorrect resets can lead to higher loss and unexpected behavior.
- **Attention Masking**: Verify that the attention masks are correctly applied. Incorrect masking can cause the model to attend to irrelevant tokens, leading to increased loss.
- **Data Packing Implementation**: Double-check the data packing implementation. Ensure that documents are packed correctly and that the model receives the expected input format.

We've used tools like **Weights & Biases** for tracking experiments and **TensorBoard** for visualizing attention patterns to debug similar issues. Additionally, consider adding logging to output the packed data and attention masks at each step. This can help identify where the discrepancy arises.

**Comment by Ueaj-Kerman** [comment]
> Per your results, "5 docs cp=2, ring_attn_func=varlen_llama3" has close loss to baseline. Is this not correct?

varlen isn't ring attention though

**Comment by Ueaj-Kerman** [comment]
man the bot problem on gh is so bad damn

**Comment by NanoCode012** [comment]
> varlen isn't ring attention though

https://docs.axolotl.ai/docs/sequence_parallelism.html#configuration

I think my colleague has patched it to work with packing. It's how we have it set for SP+packing.

**Comment by Hetens** [comment]
I reproduced this on current `main` (v0.18.0) and traced the mechanism. The reporter's guess in the "Possible solution" box is essentially right: `batch_ring` doesn't respect the document boundaries inside a packed sequence.

## Reproduction (axolotl 0.18.0, 2× GPU, SmolLM2-135M, seq_len 1024)

Using @Ueaj-Kerman's minimal repro, step-0 loss:

| config | step-0 loss | ratio to cp=1 |
|---|---|---|
| **multi-doc packed** (5 docs, `position_ids` reset at each boundary) | | |
| `cp=1` baseline | `3.171875` | 1.000× |
| `cp=2, ring_attn_func=batch_ring` | `5.812500` | **1.833×** |
| `cp=2, ring_attn_func=varlen_llama3` | `3.171875` | 1.000× |
| **single-doc control** (one contiguous doc, no internal boundaries) | | |
| `cp=1` baseline | `3.031250` | 1.000× |
| `cp=2, ring_attn_func=batch_ring` | `3.031250` | 1.000× |
| `cp=2, ring_attn_func=varlen_llama3` | `3.031250` | 1.000× |

![batch_ring loss inflation vs document boundaries](https://raw.githubusercontent.com/Hetens/axolotl/issue-3608-repro/.repro-artifacts/3608-batch_ring-loss.png)

This matches the original report (1.84× there, 1.83× here). The **single-doc control is bitwise-identical across all three configs** — so context parallelism itself is fine; the divergence appears *only* when a packed sequence contains multiple documents. `varlen_llama3` lands exactly on the baseline. ([wandb run with all curves and per-config ratios](https://wandb.ai/hetansh-shah95-umass-amherst/axolotl-3608/runs/0knmqfck))

The multi-doc dataset alternates a 5-doc pack (steps 0/2/4) with a 1-doc pack (steps 1/3); `batch_ring` diverges on the 5-doc pack and coincides on the 1-doc pack, which is the same effect visible within a single run.

## Mechanism

`src/axolotl/monkeypatch/ring_attn/adapters/batch.py` — the `batch_ring` forward accepts, and explicitly discards, every piece of document-boundary information:

```
L69   position_ids:  torch.Tensor | None = None      →  L94  "Not used in this implementation."
L75   cu_seq_lens_q: torch.LongTensor | None = None  →  L102 "Not used in this implementation."
L76   cu_seq_lens_k: torch.LongTensor | None = None  →  L103 "Not used in this implementation."
L77   max_length_q:  int | None = None               →  L104 "Not used in this implementation."
```

then at L137 it calls through with q/k/v and a single scalar `causal` flag:

```python
attn_output = RING_ATTN_FUNC_MAPPING[ring_attn_func](
    query_states, key_states, value_states,
    dropout_p=dropout, softmax_scale=softmax_scale, causal=causal,
    window_size=window_size, alibi_slopes=None, deterministic=deterministic,
    return_attn_probs=False, group=process_group,
)
```

`RING_ATTN_FUNC_MAPPING[BATCH_RING]` is `ring_flash_attn_func` — the **batch** API, which has no varlen concept. There's no channel through which a document boundary could be expressed, so the ring pass treats the whole packed sequence as one contiguous document and rotates K/V straight across every internal boundary. Queries from document B attend to keys from document A.

That predicts the observations exactly: one document → no boundary to cross → bitwise-identical; five documents → each attends to all preceding docs → loss inflated, and inflated *upward* (3.17 → 5.81) because tokens are conditioned on irrelevant prior documents. The `varlen_llama3` branch (`ring_attn/patch.py:186`) delegates to the varlen-aware `hf_adapter`, which derives `cu_seqlens` and keeps ring rotation within each document — hence it comes out clean.

## Why the config lands on `batch_ring`

`src/axolotl/utils/schemas/validation.py:1748`:

```python
sample_packing = getattr(self, "sample_packing", False)
self.ring_attn_func = (
    RingAttnFunc.VARLEN_LLAMA3 if sample_packing else RingAttnFunc.BATCH_RING
)
```

@NanoCode012 — I think this is the crux of the crossed wires earlier in the thread. Your "we set to varlen if CP and sample packing" is exactly this code and is correct. But it keys on the **`sample_packing` flag** ("did *Axolotl* pack this?"), not on whether the data actually **contains** multiple documents. @Ueaj-Kerman's data arrives **pre-packed** (`position_ids` reset per doc) with `sample_packing: false`, so the flag says "not packed", the default picks `batch_ring`, and the boundary-blind path silently corrupts the loss. You were both right: `varlen_llama3` is the working configuration, and it isn't the ring path under test — it's the workaround, not the ring implementation being fixed.

## No validator covers this

`validation.py:187` already encodes the right *concept*:

```python
def check_sample_packing_without_attention(self):
    if self.sample_packing and not self.attn_decontaminates_packing:
        LOG.warning("`sample_packing` with `attn_implementation=%r` does not handle "
                    "cross-sample decontamination. ...")
```

but it keys on `attn_implementation`, not `ring_attn_func`, so `batch_ring` never reaches it. Nothing pairs `ring_attn_func: batch_ring` with the presence of document boundaries.

## Possible directions

1. **Detect resets in `position_ids`** and route to `varlen_llama3` regardless of the `sample_packing` flag. Correct-by-default, but changes behaviour for existing configs.
2. **Extend the decontamination check** to treat `batch_ring` as a non-decontaminating backend and raise (not warn) when boundaries are present. Smallest diff, in-idiom with the existing validator, and turns silent corruption into a loud failure.
3. Both — (2) now as the safety net, (1) as the ergonomic fix.

I'd lean toward (2) first: an error beats silently-degraded training, and it doesn't require agreeing on auto-detection semantics. Happy to open a PR for whichever direction you prefer — and to hold off if someone's already on it.

<sub>Repro details: axolotl 0.18.0, transformers 5.14.1, ring-flash-attn 0.1.8, flash-attn 2.8.3, 2× A16, bf16. `NCCL_P2P_DISABLE=1` (no-NVLink node) and a transformers-5.x symbol-location shim were needed to run — both are environment plumbing and don't touch the ring-attention or loss path. Seeds pinned; runs are deterministic across invocations.</sub>
